WebGLPaintTask refactoring

emilio opened the pull request:
I refactored the current implementation, take a look and review it if necesary.

My main concern is that it won't fail nicely when there's no available WebGL context (mainly everything non-linux right now).

The idea would be to return a `Result` from `WebGLPaintTask#new` and `WebGLPaintTask#start`, and check it from `WebGLContext#new_inherited`, and `getContext`, but since `WebGLPaintTask` is not (_safely_) sendable, I haven't done nothing yet, at least without discussing it.

I think it's way cleaner right now anyways :P

Ms2ger committed:
Make ScriptTask::js_context immutable.

Since we drop the ScriptTask almost immediately after clearing the field,
there's little reason to do it manually.

emilio commented:
By the way, I made the pull request here to avoid having multiple overlapping PR's in Servo.

It's ok, right?

emilio commented:
ping @dmarcos :)

jdm commented:
Since the PR to Servo is about to merge, it probably just makes sense to reopen this PR against the servo repository instead.

emilio committed:
WebGL context acceleration support and error detection

Commit:
Auto merge of #5752 - Ms2ger:cx-immutable, r=jdm

Since we drop the ScriptTask almost immediately after clearing the field,
there's little reason to do it manually.

<!-- Reviewable:start -->
[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/servo/servo/5752)
<!-- Reviewable:end -->

Adenilson committed:
Skip computing fragment position and return early if visibility property
is not set.

Commit:
Auto merge of #5759 - Adenilson:visibleReturn01, r=pcwalton



<!-- Reviewable:start -->
[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/servo/servo/5759)
<!-- Reviewable:end -->

Commit:
Auto merge of #5763 - mbrubeck:trace-layout-readme, r=glennw

r? @glennw

<!-- Reviewable:start -->
[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/servo/servo/5763)
<!-- Reviewable:end -->

mbrubeck committed:
Update layout viewer docs

jdm committed:
Disable intermittently-passing test pending further investigation. #5764

Commit:
Auto merge of #5765 - jdm:intermittentrestore, r=jdm



<!-- Reviewable:start -->
[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/servo/servo/5765)
<!-- Reviewable:end -->

Commit:
Auto merge of #5652 - dmarcos:webgl, r=jdm

@jdm @ecoal95 I'm working on making VR happen in the Browser and I want to bring to Servo the [webVR APIs](https://github.com/MozVR/webvr-spec/blob/master/webvr.idl) we already have in Gecko. Before anything happens we need a working implementation of WebGL (and also the [fullscreen API](https://fullscreen.spec.whatwg.org/)). My implementation is very basic and probably naive (I just recently started to contribute to Servo). My patch is just a starting point: 

- It only implements ```clearColor``` and ```clear``` methods of the [WebGL spec](https://www.khronos.org/registry/webgl/specs/latest/).
- It uses the readback strategy that ```canvasRenderingContext2D``` is using (The webgl task paints stuff independently on it's own buffer and the compositor task request the pixels back to the webgl task when it needs them) I'm sure there are much better ways to handle this. Latency and FPS are critical in VR so we have to figure out the fastest way to push pixels to the screen. I've read something about layerizing the canvas but I'm still not sure what that even means :)
- There's an included test you can try ```./mach run tests/ref/webgl-context/clearcolor.html``` 

@ecoal95 I know you'll be working on this for the next three months. With a foundation in place we will be able to make quick progress in parallel. This is exciting!

<!-- Reviewable:start -->
[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/servo/servo/5652)
<!-- Reviewable:end -->

emilio committed:
Updated master

emilio committed:
Merge

emilio committed:
remove trash
